packages/grid/x-grid-modules/src/components/checkbox-renderer.tsx

```diff
@@ -27,6 +27,15 @@ export const HeaderCheckbox: React.FC<ColParams> = React.memo(({ api }) => {
     return api.onSelectionChange(selectionChange);
   }, [api, selectionChange]);
 
+  React.useEffect(() => {
+    const selectedRows = api.getSelectedRows();
+    const allRowsCount = api.getAllRowIds().length;
+    const isCurrentChecked = selectedRows.length === allRowsCount;
+    const isCurrentIndeterminate = selectedRows.length > 0 && selectedRows.length !== selectedRows;
+    setChecked(isCurrentChecked);
+    setIndeterminate(isCurrentIndeterminate);
```

Comment:
Why have the logic in an effect? Would it be simpler to pull the value from the state at render time, with the context? 

How will we handle the controllable API? #246: https://next.material-ui.com/components/data-grid/selection/#controlled-selection

Reply:
All these derived states should be computable at render time which the naming of the methods indicated. E.g. `get*` implies to me that there are now side-effects which means it's safe to call at render time. Then `getDerivedStateFromProps` would apply which has a documented pattern for hooks: https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops